cocotb/drivers/amba.py

```diff
@@ -220,98 +221,176 @@ class AXI4Slave(BusDriver):
 
         "RREADY",  "RVALID",  "RDATA",   "RLAST",   # Read response channel
 
-        "AWREADY", "AWADDR",  "AWVALID",            # Write address channel
+        "AWREADY", "AWADDR",  "AWVALID",  "AWID",   # Write address channel
         "AWPROT",  "AWSIZE",  "AWBURST", "AWLEN",
 
-        "WREADY",  "WVALID",  "WDATA",
+        "WREADY",  "WVALID",  "WDATA", "WLAST", "WSTRB", "WID",
+
+        "BVALID",  "BREADY",  "BRESP", "BID",
 
     ]
 
     # Not currently supported by this driver
     _optional_signals = [
-        "WLAST",   "WSTRB",
-        "BVALID",  "BREADY",  "BRESP",   "RRESP",
         "RCOUNT",  "WCOUNT",  "RACOUNT", "WACOUNT",
         "ARLOCK",  "AWLOCK",  "ARCACHE", "AWCACHE",
-        "ARQOS",   "AWQOS",   "ARID",    "AWID",
-        "BID",     "RID",     "WID"
+        "ARQOS",   "AWQOS",   "ARID",
+        "RID",     "RRESP",
     ]
 
-    def __init__(self, entity, name, clock, memory, callback=None, event=None,
-                 big_endian=False):
+    def __init__(self, entity, name, clock, memory, big_endian=False,
+                awready_delay=-1, wready_delay=1, bresp_delay=1):
 
         BusDriver.__init__(self, entity, name, clock)
         self.clock = clock
 
         self.big_endian = big_endian
-        self.bus.ARREADY.setimmediatevalue(1)
         self.bus.RVALID.setimmediatevalue(0)
-        self.bus.RLAST.setimmediatevalue(0)
-        self.bus.AWREADY.setimmediatevalue(1)
+        self.bus.BVALID.setimmediatevalue(0)
         self._memory = memory
 
-        self.write_address_busy = Lock("%s_wabusy" % name)
-        self.read_address_busy = Lock("%s_rabusy" % name)
-        self.write_data_busy = Lock("%s_wbusy" % name)
+        self.awready_delay = awready_delay
+        self.wready_delay = wready_delay
+        assert bresp_delay > 0, "bresp_delay < 1 not supported"
+        self.bresp_delay = bresp_delay
 
         cocotb.fork(self._read_data())
-        cocotb.fork(self._write_data())
+        cocotb.fork(self._write_transaction())
 
     def _size_to_bytes_in_beat(self, AxSIZE):
         if AxSIZE < 7:
             return 2 ** AxSIZE
         return None
 
+    @cocotb.coroutine
+    def _saxi_handshake(self, clock, ready, valid, delay=0):
+        """ Wait for a valid handshake on the supplied AXI signals
+
+        delay specifies the clock cycles to wait before asserting ready
+        after the master asserts valid.
+            delay < 0: ready is asserted before valid.
+            delay = 0: ready asserted combinatorially with valid.
+            delay > 0: wait 'delay' clock cycles after valid to assert ready.
+        """
+        ready <= (0, 1)[delay < 0]
```

Comment:
More pythonic would be `1 if delay < 0 else 0`, or perhaps just `int(delay < 0)`

Reply:
Oddly enough `ready <= 1 if delay < 0 else 0` doesn't work as I expected... at least not the same as ready <= `int(delay < 0)`

Reply:
It needs extra parentheses, `ready <= (... if ...  else ...)`, because = and <= have different precedence